src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Display,
+    str::FromStr,
+};
+
+use cosmian_crypto_core::{bytes_ser_de::Serializable, CsRng};
+use rand::SeedableRng;
+use tiny_keccak::{Hasher, Kmac};
+
+use crate::{
+    core::{
+        EncryptedTable, FindexCallbacks, FindexSearch, FindexUpsert, IndexedValue, KeyingMaterial,
+        Keyword, Uid, UpsertData,
+    },
+    error::FindexErr,
+    interfaces::{
+        generic_parameters::{
+            DemScheme, KmacKey, BLOCK_LENGTH, DEM_KEY_LENGTH, KMAC_KEY_LENGTH, KWI_LENGTH,
+            MASTER_KEY_LENGTH, TABLE_WIDTH, UID_LENGTH,
+        },
+        ser_de::serialize_set,
+    },
+};
+
+pub(crate) struct FindexCloud {
+    pub(crate) token: Token,
+    pub(crate) base_url: Option<String>,
+}
+
+pub(crate) struct Token {
+    index_id: String,
```

Comment:
I don't know what is this `index_id`. What does it represent? Why do we need it? What does it look like?

I understand it is a 4-byte string from the `split` line 74. But that's about all I've got.

Can't you use a fixed-size array of bytes to represent it ?

Reply:
It's a 5 characters ID representing the index inside the database. I will document the token format.

Reply:
@Manuthor Is 566f115 more clear?

Reply:
thanks for the additional comments.
Is this `index_id` sensitive? is it related to the Findex algorithm? Isn't it a problem this size limit on 5 characters?

Reply:
another remark: why can't we have an another master key (an authentication_master_key) which will be derived in 4 authentication keys? (it will reduce the token size)

Reply:
`index_id` is not sensitive, it can be public it only serve to identify the index and obfuscate the number of index inside our database. See my new comment in 31f6a39

Reply:
> another remark: why can't we have an another master key (an authentication_master_key) which will be derived in 4 authentication keys? (it will reduce the token size)

I think I already propose something similar to @chloehebant but it wasn't good security wise… But what you propose is a little bit different, maybe it's fine? I will ask her.

Reply:
> > another remark: why can't we have an another master key (an authentication_master_key) which will be derived in 4 authentication keys? (it will reduce the token size)
> 
> I think I already propose something similar to @chloehebant but it wasn't good security wise… But what you propose is a little bit different, maybe it's fine? I will ask her.

It's not working because with this system we cannot authorize a client to do one callback and not another?

Reply:
Indeed, if we have only one key used to derive the others, there is no way to prevent a given client from deriving any callback authentication key.

Reply:
The maximum we can do to preserve rights is to use one seed (what we call "key" in this discussion is a "seed" from a cryptographic point of view since it is used to derive other keys) per right. A seed can then be used to derive the callback authorization keys associated to the corresponding right.